client/rpc/server/server.go

```diff
@@ -0,0 +1,975 @@
+}
+
+// rpcDeserializetionError is a convenience function to convert a
+// deserialization error to an RPC error with the appropriate code set.
+func rpcDeserializationError(fmtStr string, args ...interface{}) *dcrjson.RPCError {
+	return dcrjson.NewRPCError(dcrjson.ErrRPCDeserialization,
+		fmt.Sprintf(fmtStr, args...))
+}
+
+// rpcRuleError is a convenience function to convert a
+// rule error to an RPC error with the appropriate code set.
+func rpcRuleError(fmtStr string, args ...interface{}) *dcrjson.RPCError {
+	return dcrjson.NewRPCError(dcrjson.ErrRPCMisc,
+		fmt.Sprintf(fmtStr, args...))
+}
+
+// rpcAddressKeyError is a convenience function to convert an address/key error to
+// an RPC error with the appropriate code set.  It also logs the error to the
+// RPC server subsystem since internal errors really should not occur.  The
+// context parameter is only used in the log message and may be empty if it's
+// not needed.
+func rpcAddressKeyError(fmtStr string, args ...interface{}) *dcrjson.RPCError {
+	return dcrjson.NewRPCError(dcrjson.ErrRPCInvalidAddressOrKey,
+		fmt.Sprintf(fmtStr, args...))
+}
+
+// rpcDecodeHexError is a convenience function for returning a nicely formatted
+// RPC error which indicates the provided hex string failed to decode.
+func rpcDecodeHexError(gotHex string) *dcrjson.RPCError {
+	return dcrjson.NewRPCError(dcrjson.ErrRPCDecodeHexString,
+		fmt.Sprintf("Argument must be hexadecimal string (not %q)",
+			gotHex))
+}
+
+// rpcMiscError is a convenience function for returning a nicely formatted RPC
+// error which indicates there is an unquantifiable error.  Use this sparingly;
+// misc return codes are a cop out.
+func rpcMiscError(message string) *dcrjson.RPCError {
+	return dcrjson.NewRPCError(dcrjson.ErrRPCMisc, message)
+}
+
+// handleHelp implements the help command.
+func handleHelp(_ context.Context, s *server, cmd interface{}) (interface{}, error) {
+	return "help", nil
+}
+
+// handleVersion implements the version command.
+func handleVersion(_ context.Context, s *server, cmd interface{}) (interface{}, error) {
+	result := VersionResult{
+		Major: jsonrpcSemverMajor,
+		Minor: jsonrpcSemverMinor,
+		Patch: jsonrpcSemverPatch,
+	}
+	return result, nil
+}
+
+// server provides a concurrent safe RPC server.
+type server struct {
+	started      int32
+	shutdown     int32
+	cfg          Config
+	authsha      [sha256.Size]byte
+	limitauthsha [sha256.Size]byte
+	ntfnMgr      *wsNotificationManager
+	numClients   int32
+	statusLines  map[int]string
+	statusLock   sync.RWMutex
+	wg           sync.WaitGroup
+	//helpCacher             *helpCacher
+	requestProcessShutdown chan struct{}
+}
+
+// httpStatusLine returns a response Status-Line (RFC 2616 Section 6.1) for the
+// given request and response status code.  This function was lifted and
+// adapted from the standard library HTTP server code since it's not exported.
+func (s *server) httpStatusLine(req *http.Request, code int) string {
```

Comment:
It seems to me that `httpStatusLine` and `writeHTTPResponseHeaders` are overkill and we just need `(*http.ResponseWriter).WriteHeader` and `(http.Header).Set`.